File: src/main/java/de/istkorrekt/gameoflife/app/cli/CliApplication.java
package de.istkorrekt.gameoflife.app.cli;

import de.istkorrekt.gameoflife.Location;
import de.istkorrekt.gameoflife.Matchfield;
import de.istkorrekt.gameoflife.MatchfieldFactory;
import de.istkorrekt.gameoflife.Size;
import de.istkorrekt.gameoflife.util.LocationParser;

import java.time.Duration;
import java.time.Instant;
import java.util.Collection;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.logging.Logger;
import java.util.stream.Stream;

public class CliApplication {

    private static final Logger LOG = Logger.getLogger("game");

    public static void main(String[] args) {
        Collection<Location> locations = LocationParser.parse(args[0]);
        int size = Integer.parseInt(args[1]);
        Matchfield matchfield = new MatchfieldFactory().createWithAlive(locations, Size.of(size));

        if (args.length >= 3) {
            int rounds = Integer.parseInt(args[2]);
            autoPlay(matchfield, rounds);
        } else {
            LOG.info(matchfield.toString());
        }
    }

    private static void autoPlay(Matchfield initialMatchfield, int rounds) {
        Instant start = Instant.now();

        AtomicInteger roundsCounter = new AtomicInteger();
        Stream.iterate(initialMatchfield, Matchfield::nextGeneration)
                .limit(rounds + 1)
                .peek(game -> LOG.info("Round " + roundsCounter.getAndIncrement() + "\n" + game))
                .reduce((first, second) -> second)
                .orElseThrow();

        LOG.info("\nGame finished within " + Duration.between(start, Instant.now()) + " after " + rounds + " rounds.");
    }
}

File: src/main/kotlin/de/istkorrekt/gameoflife/app/cli/CliApplication.kt
package de.istkorrekt.gameoflife.app.cli

import de.istkorrekt.gameoflife.Location
import de.istkorrekt.gameoflife.Matchfield
import de.istkorrekt.gameoflife.MatchfieldFactory.createWithAlive
import de.istkorrekt.gameoflife.Size
import de.istkorrekt.gameoflife.util.parse
import java.util.logging.Logger
import kotlin.time.ExperimentalTime
import kotlin.time.measureTime

val logger: Logger = Logger.getGlobal()

fun main(args: Array<String>) {
    val locations: Collection<Location> = parse(args[0])
    val size = args[1].toInt()
    val matchfield = createWithAlive(Size(size), locations)

    if (args.size >= 3) {
        val rounds = args[2].toInt()
        autoPlay(matchfield, rounds)
    } else {
        logger.info(matchfield.toString())
    }
}

@OptIn(ExperimentalTime::class)
fun autoPlay(initialMatchfield: Matchfield, rounds: Int) {
    val duration = measureTime {
        MatchfieldIterator(initialMatchfield, rounds).forEach {
            logger.info(it.toString())
        }
    }
    logger.info("Game finished within $duration.d after $rounds rounds.")
}

private class MatchfieldIterator(var current: Matchfield, val rounds: Int) : Iterator<Matchfield> {
    var currentRound = 1

    override fun hasNext(): Boolean {
        return currentRound < rounds
    }

    override fun next(): Matchfield {
        check(hasNext())
        current = current.nextGeneration()
        currentRound++
        return current
    }
}